### PR TITLE
Hack out effects support for old solver

### PR DESCRIPTION
**Pull request by compiler-errors**
Opening this for vibes ✨ 

Turns out that a basic, somewhat incomplete implementation of host effects is achievable in the old trait solver pretty easily. This should be sufficient for us to use in the standard library itself.

Regarding incompleteness, maybe we should always treat host predicates as ambiguous in intercrate mode (at least in the old solver) to avoid any worries about accidental impl overlap or something.

r? @lcnr cc @fee1-dead

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/effects.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                data.skip_binder().trait_ref.args,
+            ) {
+                continue;
+            }
+
+            let is_match = infcx.probe(|_| match_candidate(infcx, obligation, data).is_ok());
+
+            if is_match {
+                if candidate.is_some() {
+                } else {
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/effects.rs:85]: nyaaa?

either add a comment or modify the impl

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/effects.rs:85]: yeah this impl is scuffed lol

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/effects.rs:85]: This should've been a `return Err(EvaluationFailure::Ambiguous)`, so we bail on ambiguity if there are 2 applicable w.c.

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 2]
```diff
@@ -645,11 +645,19 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
...
                 }
 
-                ty::PredicateKind::Clause(ty::ClauseKind::HostEffect(..)) => {
-                    // FIXME(effects): It should be relatively straightforward to implement
-                    // old trait solver support for `HostEffect` bounds; or at least basic
-                    // support for them.
-                    todo!()
+                ty::PredicateKind::Clause(ty::ClauseKind::HostEffect(data)) => {
+                    self.infcx.enter_forall(bound_predicate.rebind(data), |data| {
+                        match effects::evaluate_host_effect_obligation(
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:650]: pretty sure that can cause undetected overflow :thinking: too tired to explain my thoughts here via comment for today, can do so in sync or maybe you're able to come up with a test yourself

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/select/mod.rs:650]: Yeah, though it's the same way that we can encounter overflow in projection afaict.

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/select/mod.rs:650]: This is just a consequence of not tracking effect goals (or projections) on the trait stack.

**Review by fee1-dead — COMMENTED** [review]
The test changes appear correct to me. For the solver parts I did not see anything else than what lcnr commented on (in my limited understanding of the trait solver infra)

File: compiler/rustc_trait_selection/src/traits/fulfill.rs [hunk 3]
```diff
@@ -854,6 +861,40 @@ impl<'a, 'tcx> FulfillProcessor<'a, 'tcx> {
...
         }
     }
+
+    fn process_host_obligation(
+        &mut self,
+        obligation: &PredicateObligation<'tcx>,
+        host_obligation: HostEffectObligation<'tcx>,
+        stalled_on: &mut Vec<TyOrConstInferVar>,
+    ) -> ProcessResult<PendingPredicateObligation<'tcx>, FulfillmentErrorCode<'tcx>> {
+        if obligation.predicate.is_global() && !self.selcx.is_intercrate() {
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/fulfill.rs:871]: We can remove this.

File: compiler/rustc_trait_selection/src/traits/effects.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                        return Err(EvaluationFailure::NoSolution);
+                    }
+
+                    let mut nested = impl_.nested;
+                    nested.extend(
+                        tcx.const_conditions(impl_.impl_def_id)
+                            .instantiate(tcx, impl_.args)
+                            .into_iter()
+                            .map(|(trait_ref, _)| {
+                                obligation.with(
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/effects.rs:122]: Increment the depth correctly

**Review by lcnr — COMMENTED** [review]
pls do:
- incr obligation depth when recursing in evaluate host effect
- ICE when used with intercrate mode

r=me after

File: compiler/rustc_trait_selection/src/traits/effects.rs (hunk 4 above)
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/effects.rs:135]: Did this more manually. It's kinda ugly but 🤷

**Pull request merged**